ml-agents-envs/mlagents/envs/base_env.py

```diff
@@ -0,0 +1,291 @@
+
+
+class BatchedStepResult(NamedTuple):
+    """
+    Contains the data a group of similar Agents collected since the last
+    simulation step. Note that all Agents do not necessarily have new
+    information to send at each simulation step. Therefore, the ordering of
+    agents and the batch size of the BatchedStepResult are not fixed across
+    simulation steps.
+     - obs is a list of numpy arrays observations collected by the group of
+     agent. Each obs has one extra dimension compared to StepResult: the first
+     dimension of the array corresponds to the batch size of
+     the group.
+     - reward is a float vector of length batch size. Corresponds to the
+     rewards collected by each agent since the last simulation step.
+     - done is an array of booleans of length batch size. Is true if the
+     associated Agent was terminated during the last simulation step.
+     - max_step is an array of booleans of length batch size. Is true if the
+     associated Agent reached its maximum number of steps during the last
+     simulation step.
+     - agent_id is an int vector of length batch size containing unique
+     identifier for the corresponding Agent. This is used to track Agents
+     across simulation steps.
+     - action_mask is an optional list of two dimensional array of booleans.
+     Only available in multi-discrete action space type.
+     Each array corresponds to an action branch. The first dimension of each
+     array is the batch size and the second contains a mask for each action of
+     the branch. If true, the action is not available for the agent during
+     this simulation step.
+    """
+
+    obs: List[np.array]
+    reward: np.array
+    done: np.array
+    max_step: np.array
+    agent_id: np.array
+    action_mask: Optional[List[np.array]]
+
+    def get_agent_step_result(self, agent_id: int) -> StepResult:
+        """
+        returns the step result for a specific agent.
+        :param agent_id: The id of the agent
+        :returns: obs, reward, done, agent_id and optional action mask for a
+        specific agent
+        """
+        try:
+            agent_index = np.where(self.agent_id == agent_id)[0][0]
+        except IndexError as ie:
+            raise IndexError(
+                "agent_id {} is not present in the BatchedStepResult".format(agent_id)
+            ) from ie
+        agent_obs = []
+        for batched_obs in self.obs:
+            agent_obs.append(batched_obs[agent_index])
+        agent_mask = None
+        if self.action_mask is not None:
+            agent_mask = []
+            for mask in self.action_mask:
+                agent_mask.append(mask[0])
+        return StepResult(
+            obs=agent_obs,
+            reward=self.reward[agent_index],
+            done=self.done[agent_index],
+            max_step=self.max_step[agent_index],
+            agent_id=agent_id,
+            action_mask=agent_mask,
+        )
+
+    @staticmethod
+    def empty(spec):
```

Comment:
```suggestion
    def empty(spec: AgentGroupSpec) -> "BatchedStepResult":
```
(you need the quotes since this is basically a forward declaration)